### ISSUE TITLE
parsing problem due to parentheses, again

### ISSUE THREAD

**Issue by grosu** [kframework/k-legacy#179]
```
fslwork$ cat try.k
module TRY
  syntax Exp ::= Id
         | "(" Exp ")"                       [bracket]
         | Exp "(" Exps ")"                  [strict, prefer]
         > Exp Exp                           [strict, left]
         > Exp "*" Exp
  syntax Exps ::= List{Exp,","}
endmodule
fslwork$ krun --parser "kast -e" "x * f x"
<k>
    x * (f x)
</k>
fslwork$ krun --parser "kast -e" "x * f(x)"
[Warning] Parser: Parsing ambiguity. Arbitrarily choosing the first.
1: Exp ::= Exp "*" Exp 
   x * (f ( (x, .Exps) ))
2: Exp ::= Exp "*" Exp 
   x * (f (x))
    File: x * f(x)
    Location: (1,1,1,9)
    Compilation Phase: Ambiguity filter

<k>
    x * (f ( x ))
</k>
fslwork$ 
```

The reported ambiguity is incorrect, because both productions are the same.  The ambiguity is probably below, on "f(x)", but there should be no ambiguity there either, because the production Exp "(" Exps ")" has priority.

Radu, in case this is a simple fix, I would be grateful if you can push it asap.  Fixing this bug would allow me to have a much nicer FUN language semantics.

**Comment by radumereuta** [comment]
The problem is that I'm using a special class called kil.Bracket which instantiates a bracket from the AST.
When I'm applying the priority filter, I'm avoiding this object.

There are two possible solutions for this:
1) A bit more hackish, but faster: take into consideration the Bracket class when disambiguating.
2) A bit cleaner: remove the Bracket class and use the generic TermCons class.

The problem with the second solution is backwards compatibility. I did a quick search for the Bracket class and it is being used in several places in the unparser.
I would like to implement the second version because it's cleaner and more correct. It simplifies the code instead of making it more complex, but I'm too tired to do it tonight, and I don't want to break the build.

If this is really urgent, I can try it later when I get home, if not, I would like to do it first thing tomorrow.

**Comment by grosu** [comment]
Tomorrow is fine, I found a way to go around it.  It is important to pick the clean solutions.

Grigore

**Comment by dwightguth** [comment]
Did this ever get addressed?

**Comment by radumereuta** [comment]
I don't think so. From what I remember, I wanted to remove the Bracket class from KIL, but it was used in the unparser and I didn't know how to modify the code.

I think it's better to delay it for the new version of the parser.
I've assigned myself so I know to revisit it after a while.
